### PR TITLE
Fix a broken debug assertion that kicked in whenever building Fsharp.Core

### PR DESCRIPTION
**Pull request by T-Gro**
When Fsharp.Core was being built with Debug. the assertion 

`assert(vspec.InlineInfo = inlineFlag)`

at the very end of CheckExpressions was broken.
This was in main.

Reason is, the [<MethodImpl.NoInlining>] was not reflected, and only the newly introduced (F# specific) attrib_NoCompilerInliningAttribute  was.

The bigger infrastructural issue is that this slipped trough in the first place.
I think we should build at least some projects also in Debug so that assertions get exercised.

### CODE REVIEW

**Comment by T-Gro** [comment]
@kerams  : Can you please check if this does not violate some older agreements about what  [<MethodImpl.NoInlining>] or should not do?

**Review by KevinRansom — DISMISSED** [review]

**Review by abonie — APPROVED** [review]

File: src/Compiler/Checking/CheckExpressions.fs [hunk 1]
```diff
@@ -12042,6 +12042,12 @@ let TcAndPublishValSpec (cenv: cenv, env, containerInfo: ContainerInfo, declKind
     let attrs = TcAttributes cenv env attrTgt synAttrs
     let newOk = if canInferTypars then NewTyparsOK else NoNewTypars
     let hasNoCompilerInliningAttribute = HasFSharpAttribute g g.attrib_NoCompilerInliningAttribute attrs
+    let hasMethodImplNoInlining = 
+            match TryFindFSharpAttribute g g.attrib_MethodImplAttribute attrs with
+            // NO_INLINING = 8
+            | Some (Attrib(_, _, [ AttribInt32Arg flags ], _, _, _, _)) -> (flags &&& 0x8) <> 0x0
+            | _ -> false
+    let forceDoNotInline = hasNoCompilerInliningAttribute || hasMethodImplNoInlining
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckExpressions.fs:12050]: I'm not entirely sure it's correct or am I missing something?

`attrib_MethodImplAttribute` is a hint for JIT only, I am not entirely sure we need to be looking at it.

Presense of `MethodImplAttribute.NoInline` does not mean F# compiler shouldn't inline it, does it?

**Review by vzarytovskii — CHANGES_REQUESTED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Review by abonie — APPROVED** [review]

File: src/Compiler/Checking/CheckExpressions.fs [hunk 2]
```diff
@@ -2205,28 +2205,31 @@ module GeneralizationHelpers =
...
         elif isInline then
-            ValInline.Always 
+            ValInline.Always, ignore
         else
-            ValInline.Optional
+            ValInline.Optional, ignore
 
     if isInline && (inlineFlag <> ValInline.Always) then 
-        errorR(Error(FSComp.SR.tcThisValueMayNotBeInlined(), m))
+        reportIncorrectInlineKeywordUsage (Error(FSComp.SR.tcThisValueMayNotBeInlined(), m))
```
Comment (abonie) [inline src/Compiler/Checking/CheckExpressions.fs:2232]: I think at this point `(inlineFlag <> ValInline.Always)` above is no longer needed. Overall this became hard to read now IMO, but no strong feelings, feel free to resolve this comment if you disagree.

**Review by abonie — DISMISSED** [review]

File: src/Compiler/Checking/CheckExpressions.fs (hunk 2 above)
Comment (T-Gro) [reply src/Compiler/Checking/CheckExpressions.fs:2232]: It is not equivalent, as it processes the HasMethodImplNoInliningAttribute again for a second time.
Also, if something would be e.g. both isCtorOrAbstractSlot AND HasMethodImplNoInliningAttribute  => should be an error, not a warning.

I had something similar to what you have (i.e. repeating the if-then logic in the diagnosticla part), but then reverted it to:
- call the functions at max once
- keep the property of short circuiting ( via || and via if-then) to not read an attribute when it can be avoided.

Comment (abonie) [reply src/Compiler/Checking/CheckExpressions.fs:2232]: You're right, it is not the same, it's kinda hard to follow. But as I said, I don't feel very strongly about it either way

File: src/Compiler/Checking/CheckExpressions.fs [hunk 3]
```diff
@@ -1209,6 +1209,12 @@ let CheckRequiredProperties (g:TcGlobals) (env: TcEnv) (cenv: TcFileState) (minf
                 let details = NicePrint.multiLineStringOfPropInfos g cenv.amap mMethExpr env.DisplayEnv missingProps
                 errorR(Error(FSComp.SR.tcMissingRequiredMembers details, mMethExpr))
 
+let private HasMethodImplNoInliningAttribute g attrs = 
+            match TryFindFSharpAttribute g g.attrib_MethodImplAttribute attrs with
```
Comment (psfinaki) [inline src/Compiler/Checking/CheckExpressions.fs:1213]: Seems like extra spacing here but doesn't matter

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/MethodImplAttribute/MethodImplAttribute.fs [hunk 4]
```diff
@@ -33,6 +33,20 @@ module MethodImplAttribute =
     let ``NoInlining_fs`` compilation =
         compilation
         |> verifyCompilation
+     
+    [<Theory; Directory(__SOURCE_DIRECTORY__, Includes=[|"MethodImplAttribute.NoInlining_InlineKeyword.fs"|])>]
+    let ``NoInlining_fs with inline keyword => should warn in preview version`` compilation =
+        compilation
+        |> withLangVersionPreview
+        |> typecheck
+        |> withSingleDiagnostic (Warning 3151, Line 2, Col 1, Line 3, Col 38, "This member, function or value declaration may not be declared 'inline'")
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/EmittedIL/MethodImplAttribute/MethodImplAttribute.fs:42]: Maybe you could have just "declaration".

**Review by psfinaki — APPROVED** [review]

**Review by psfinaki — APPROVED** [review]
